lib/GameConstants.cpp

```diff
@@ -160,3 +162,30 @@ std::string ETerrainType::toString() const
 	ss << *this;
 	return ss.str();
 }
+
+std::ostream & operator<<(std::ostream & os, const EPathfindingLayer actionType)
+{
+	static const std::map<EPathfindingLayer::EEPathfindingLayer, std::string> pathfinderLayerToString =
+	{
+	#define DEFINE_ELEMENT(element) {EPathfindingLayer::element, #element}
+		DEFINE_ELEMENT(WRONG),
+		DEFINE_ELEMENT(AUTO),
+		DEFINE_ELEMENT(LAND),
+		DEFINE_ELEMENT(SAIL),
+		DEFINE_ELEMENT(WATER),
+		DEFINE_ELEMENT(AIR),
+		DEFINE_ELEMENT(NUM_LAYERS)
+
+	};
+
+	auto it = pathfinderLayerToString.find(actionType.num);
+	if (it == pathfinderLayerToString.end()) return os << "<Unknown type>";
+	else return os << it->second;
+}
+
+std::string EPathfindingLayer::toString() const
+{
+	std::stringstream ss;
+	ss << *this;
+	return ss.str();
+}

```

Comment:
I just take ETerrainType as base because it's looks to suite my needs for development and debugging. It's experimental code as first I need write all pathfinding rules down and workaround edge cases for both findinging and validating path. Once I finish with it and make sure it's working I'll start refactoring to make it more efficient.

Feedback still really appreciated as otherwise I can miss something else.